controllers/storagecluster/cephnfs.go

```diff
@@ -0,0 +1,346 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ocsCephNFS struct{}
+
+// newCephNFSInstance returns the cephNFS instance that should be created
+// on first run.
+func (r *StorageClusterReconciler) newCephNFSInstance(initData *ocsv1.StorageCluster) (*cephv1.CephNFS, error) {
+	obj := &cephv1.CephNFS{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateNameForCephNFS(initData),
+			Namespace: initData.Namespace,
+		},
+		Spec: cephv1.NFSGaneshaSpec{
+			Server: cephv1.GaneshaServerSpec{
+				Active:    1,
+				Placement: getPlacement(initData, "nfs"),
+				Resources: defaults.GetDaemonResources("nfs", initData.Spec.Resources),
+				// set PriorityClassName for the NFS pods
+				PriorityClassName: openshiftUserCritical,
+			},
+		},
+	}
+	err := controllerutil.SetControllerReference(initData, obj, r.Scheme)
+	if err != nil {
+		r.Log.Error(err, "Unable to set Controller Reference for CephNFS.", "CephNFS", klog.KRef(obj.Namespace, obj.Name))
+		return nil, err
+	}
+	return obj, nil
+}
+
+// ensureCreated ensures that cephNFS resource exist in the desired state.
+func (obj *ocsCephNFS) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if instance.Spec.NFS == nil || !instance.Spec.NFS.Enable {
+		return reconcile.Result{}, nil
+	}
+
+	cephNFS, err := r.newCephNFSInstance(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	existingCephNFS := cephv1.CephNFS{}
+	ctxTODO := context.TODO()
+	err = r.Client.Get(ctxTODO, types.NamespacedName{Name: cephNFS.Name, Namespace: cephNFS.Namespace}, &existingCephNFS)
+	switch {
+	case err == nil:
+		if existingCephNFS.DeletionTimestamp != nil {
+			r.Log.Info("Unable to restore CephNFS because it is marked for deletion.", "CephNFS", klog.KRef(existingCephNFS.Namespace, existingCephNFS.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to restore initialization object %q because it is marked for deletion", existingCephNFS.Name)
+		}
+
+		r.Log.Info("Restoring original CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+		existingCephNFS.ObjectMeta.OwnerReferences = cephNFS.ObjectMeta.OwnerReferences
+		existingCephNFS.Spec = cephNFS.Spec
+		err = r.Client.Update(ctxTODO, &existingCephNFS)
+		if err != nil {
+			r.Log.Error(err, "Unable to update CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+			return reconcile.Result{}, err
+		}
+	case errors.IsNotFound(err):
+		r.Log.Info("Creating CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+		err = r.Client.Create(ctxTODO, cephNFS)
+		if err != nil {
+			r.Log.Error(err, "Unable to create CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+			return reconcile.Result{}, err
+		}
+	}
```

Comment:
Where is the case for handling errors? Currently this handles `err == nil` and `err != nil && err.IsNotFound(err)`, but it doesn't handle `err != nil && !errors.IsNotFound(err)` which is the actual error case.

(@jarrpa is this intentional? I also see this left out of the CephFilesystem controller)

I think the easiest way to handle this is like the flow in `ensureDeleted` below that looks like:
```go
if err != nil {
	if errors.IsNotFound(err) {
		// Create new
		r.Log.Info("Creating CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
	}
	// Error case
}
// Non-error case
r.Log.Info("Restoring original CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
```

Reply:
added default case to handle error.